In [32]:
## load libraries
import pandas as pd
import numpy as np

In [33]:
## load data
## we will be using the titanic dataset

# Create URL
### url = 'https://raw.githubusercontent.com/chrisalbon/simulated_datasets/master/titanic.csv'


## load the data
### data_frame = pd.read_csv(url)

### Write the data to the local disk


## data_frame.to_csv('dataframe.csv')

## load the data from the local disk
dataframe = pd.read_csv('dataframe.csv', index_col="Unnamed: 0")
dataframe.head()

Name PClass    Age     Sex  \
0                   Allen, Miss Elisabeth Walton    1st  29.00  female   
1                    Allison, Miss Helen Loraine    1st   2.00  female   
2            Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                  Allison, Master Hudson Trevor    1st   0.92    male   

   Survived  SexCode  
0         1        1  
1         0        1  
2         0        0  
3         0        1  
4         1        0

In [34]:
## dataset properties
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1313 entries, 0 to 1312
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      1313 non-null   object 
 1   PClass    1313 non-null   object 
 2   Age       756 non-null    float64
 3   Sex       1313 non-null   object 
 4   Survived  1313 non-null   int64  
 5   SexCode   1313 non-null   int64  
dtypes: float64(1), int64(2), object(3)
memory usage: 71.8+ KB


>- The dataset has 6 variables and 1313 entries

### Row Selection based on Conditions

In [35]:
## Select the female passengers
dataframe[dataframe['Sex'] == 'female']

Name PClass   Age     Sex  \
0                      Allen, Miss Elisabeth Walton    1st  29.0  female   
1                       Allison, Miss Helen Loraine    1st   2.0  female   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.0  female   
6                  Andrews, Miss Kornelia Theodosia    1st  63.0  female   
8      Appleton, Mrs Edward Dale (Charlotte Lamson)    1st  58.0  female   
...                                             ...    ...   ...     ...   
1283            Vestrom, Miss Hulda Amanda Adolfina    3rd  14.0  female   
1293                              Wilkes, Mrs Ellen    3rd  45.0  female   
1304                            Yasbeck, Mrs Antoni    3rd  15.0  female   
1306                            Zabour, Miss Hileni    3rd   NaN  female   
1307                            Zabour, Miss Tamini    3rd   NaN  female   

      Survived  SexCode  
0            1        1  
1            0        1  
3            0        1  
6            1        1  
8            1        1  
...        ...      ...  
1283         0        1  
1293         1        1  
1304         1        1  
1306         0        1  
1307         0        1  

[462 rows x 6 columns]

In [36]:
## filter with more than two conditions
dataframe[(dataframe['Sex'] == 'female') & (dataframe['PClass'] == '1st')] 

Name PClass   Age     Sex  \
0                     Allen, Miss Elisabeth Walton    1st  29.0  female   
1                      Allison, Miss Helen Loraine    1st   2.0  female   
3    Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.0  female   
6                 Andrews, Miss Kornelia Theodosia    1st  63.0  female   
8     Appleton, Mrs Edward Dale (Charlotte Lamson)    1st  58.0  female   
..                                             ...    ...   ...     ...   
315                                  Pericault, Ms    1st   NaN  female   
318                            Segesser, Mlle Emma    1st   NaN  female   
319                                   Seredeca, Ms    1st   NaN  female   
320                                  Ward, Ms Anna    1st   NaN  female   
321                               Wilson, Ms Helen    1st   NaN  female   

     Survived  SexCode  
0           1        1  
1           0        1  
3           0        1  
6           1        1  
8           1        1  
..        ...      ...  
315         1        1  
318         1        1  
319         0        1  
320         0        1  
321         1        1  

[143 rows x 6 columns]

In [37]:
## filter using numeric variables
## For age greater that 55 years
dataframe[dataframe['Age'] > 55].head(5)

Name PClass   Age     Sex  \
6                Andrews, Miss Kornelia Theodosia    1st  63.0  female   
8    Appleton, Mrs Edward Dale (Charlotte Lamson)    1st  58.0  female   
9                          Artagaveytia, Mr Ramon    1st  71.0    male   
28                        Bonnell, Miss Elizabeth    1st  58.0  female   
37  Brown, Mrs John Murray (Caroline Lane Lamson)    1st  59.0  female   

    Survived  SexCode  
6          1        1  
8          1        1  
9          0        0  
28         1        1  
37         1        1

### Value Replacement

In [38]:
## Replace male with man
dataframe.replace('male', 'man').head()

Name PClass    Age     Sex  \
0                   Allen, Miss Elisabeth Walton    1st  29.00  female   
1                    Allison, Miss Helen Loraine    1st   2.00  female   
2            Allison, Mr Hudson Joshua Creighton    1st  30.00     man   
3  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                  Allison, Master Hudson Trevor    1st   0.92     man   

   Survived  SexCode  
0         1        1  
1         0        1  
2         0        0  
3         0        1  
4         1        0

In [39]:
## Replace more than one value of the same variable
dataframe.replace(['male', 'female'], ['man', 'woman'])

Name PClass    Age    Sex  \
0                      Allen, Miss Elisabeth Walton    1st  29.00  woman   
1                       Allison, Miss Helen Loraine    1st   2.00  woman   
2               Allison, Mr Hudson Joshua Creighton    1st  30.00    man   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  woman   
4                     Allison, Master Hudson Trevor    1st   0.92    man   
...                                             ...    ...    ...    ...   
1308                             Zakarian, Mr Artun    3rd  27.00    man   
1309                         Zakarian, Mr Maprieder    3rd  26.00    man   
1310                               Zenni, Mr Philip    3rd  22.00    man   
1311                               Lievens, Mr Rene    3rd  24.00    man   
1312                                 Zimmerman, Leo    3rd  29.00    man   

      Survived  SexCode  
0            1        1  
1            0        1  
2            0        0  
3            0        1  
4            1        0  
...        ...      ...  
1308         0        0  
1309         0        0  
1310         0        0  
1311         0        0  
1312         0        0  

[1313 rows x 6 columns]

In [40]:
## Renaming columns
dataframe.rename(columns={'Sex': 'Gender'})

Name PClass    Age  Gender  \
0                      Allen, Miss Elisabeth Walton    1st  29.00  female   
1                       Allison, Miss Helen Loraine    1st   2.00  female   
2               Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                     Allison, Master Hudson Trevor    1st   0.92    male   
...                                             ...    ...    ...     ...   
1308                             Zakarian, Mr Artun    3rd  27.00    male   
1309                         Zakarian, Mr Maprieder    3rd  26.00    male   
1310                               Zenni, Mr Philip    3rd  22.00    male   
1311                               Lievens, Mr Rene    3rd  24.00    male   
1312                                 Zimmerman, Leo    3rd  29.00    male   

      Survived  SexCode  
0            1        1  
1            0        1  
2            0        0  
3            0        1  
4            1        0  
...        ...      ...  
1308         0        0  
1309         0        0  
1310         0        0  
1311         0        0  
1312         0        0  

[1313 rows x 6 columns]

In [41]:
## Rename multiple ; do this by supplying a dictionary of the new-name:old-name
dataframe.rename(columns={'Sex': 'Gender', 'PClass':'Passenger_Class'})

Name Passenger_Class    Age  \
0                      Allen, Miss Elisabeth Walton             1st  29.00   
1                       Allison, Miss Helen Loraine             1st   2.00   
2               Allison, Mr Hudson Joshua Creighton             1st  30.00   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)             1st  25.00   
4                     Allison, Master Hudson Trevor             1st   0.92   
...                                             ...             ...    ...   
1308                             Zakarian, Mr Artun             3rd  27.00   
1309                         Zakarian, Mr Maprieder             3rd  26.00   
1310                               Zenni, Mr Philip             3rd  22.00   
1311                               Lievens, Mr Rene             3rd  24.00   
1312                                 Zimmerman, Leo             3rd  29.00   

      Gender  Survived  SexCode  
0     female         1        1  
1     female         0        1  
2       male         0        0  
3     female         0        1  
4       male         1        0  
...      ...       ...      ...  
1308    male         0        0  
1309    male         0        0  
1310    male         0        0  
1311    male         0        0  
1312    male         0        0  

[1313 rows x 6 columns]

In [42]:
## Total missing observations by columns
dataframe.isnull().sum()

Name          0
PClass        0
Age         557
Sex           0
Survived      0
SexCode       0
dtype: int64

### Categorical Variables

In [43]:
## Unique values
dataframe['Sex'].unique()

array(['female', 'male'], dtype=object)

In [44]:
## Value counts for the categories
dataframe['Sex'].value_counts()

male      851
female    462
Name: Sex, dtype: int64

In [45]:
## Value counts for the categories(Proportion)
dataframe['Sex'].value_counts(normalize=True)

male      0.648134
female    0.351866
Name: Sex, dtype: float64

### Handling Missing Values

In [46]:
## Check for the missing values for each column in the data frame
dataframe.isnull().sum()

Name          0
PClass        0
Age         557
Sex           0
Survived      0
SexCode       0
dtype: int64

>- There is only one variable that contains missing observations(557)

In [47]:
## replace missing observations with -999
dataframe.replace(np.NaN, -999)

Name PClass    Age     Sex  \
0                      Allen, Miss Elisabeth Walton    1st  29.00  female   
1                       Allison, Miss Helen Loraine    1st   2.00  female   
2               Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                     Allison, Master Hudson Trevor    1st   0.92    male   
...                                             ...    ...    ...     ...   
1308                             Zakarian, Mr Artun    3rd  27.00    male   
1309                         Zakarian, Mr Maprieder    3rd  26.00    male   
1310                               Zenni, Mr Philip    3rd  22.00    male   
1311                               Lievens, Mr Rene    3rd  24.00    male   
1312                                 Zimmerman, Leo    3rd  29.00    male   

      Survived  SexCode  
0            1        1  
1            0        1  
2            0        0  
3            0        1  
4            1        0  
...        ...      ...  
1308         0        0  
1309         0        0  
1310         0        0  
1311         0        0  
1312         0        0  

[1313 rows x 6 columns]

### Column deletion

In [48]:
## delete a single column by it's name
dataframe.drop('SexCode', axis=1)

Name PClass    Age     Sex  \
0                      Allen, Miss Elisabeth Walton    1st  29.00  female   
1                       Allison, Miss Helen Loraine    1st   2.00  female   
2               Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                     Allison, Master Hudson Trevor    1st   0.92    male   
...                                             ...    ...    ...     ...   
1308                             Zakarian, Mr Artun    3rd  27.00    male   
1309                         Zakarian, Mr Maprieder    3rd  26.00    male   
1310                               Zenni, Mr Philip    3rd  22.00    male   
1311                               Lievens, Mr Rene    3rd  24.00    male   
1312                                 Zimmerman, Leo    3rd  29.00    male   

      Survived  
0            1  
1            0  
2            0  
3            0  
4            1  
...        ...  
1308         0  
1309         0  
1310         0  
1311         0  
1312         0  

[1313 rows x 5 columns]

In [49]:
## drop multiple columns
dataframe.drop(['Sex', 'Name', 'SexCode'], axis=1)

PClass    Age  Survived
0       1st  29.00         1
1       1st   2.00         0
2       1st  30.00         0
3       1st  25.00         0
4       1st   0.92         1
...     ...    ...       ...
1308    3rd  27.00         0
1309    3rd  26.00         0
1310    3rd  22.00         0
1311    3rd  24.00         0
1312    3rd  29.00         0

[1313 rows x 3 columns]

### Deleting (Duplicate) Rows

>- Deleting rows is similar to filtering out rows

In [50]:
## Filter out the 2nd Class Passengers
dataframe[dataframe['PClass'] != '2nd']

Name PClass    Age     Sex  \
0                      Allen, Miss Elisabeth Walton    1st  29.00  female   
1                       Allison, Miss Helen Loraine    1st   2.00  female   
2               Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                     Allison, Master Hudson Trevor    1st   0.92    male   
...                                             ...    ...    ...     ...   
1308                             Zakarian, Mr Artun    3rd  27.00    male   
1309                         Zakarian, Mr Maprieder    3rd  26.00    male   
1310                               Zenni, Mr Philip    3rd  22.00    male   
1311                               Lievens, Mr Rene    3rd  24.00    male   
1312                                 Zimmerman, Leo    3rd  29.00    male   

      Survived  SexCode  
0            1        1  
1            0        1  
2            0        0  
3            0        1  
4            1        0  
...        ...      ...  
1308         0        0  
1309         0        0  
1310         0        0  
1311         0        0  
1312         0        0  

[1034 rows x 6 columns]

In [51]:
## Dropping duplicate rows
dataframe.drop_duplicates()

Name PClass    Age     Sex  \
0                      Allen, Miss Elisabeth Walton    1st  29.00  female   
1                       Allison, Miss Helen Loraine    1st   2.00  female   
2               Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                     Allison, Master Hudson Trevor    1st   0.92    male   
...                                             ...    ...    ...     ...   
1308                             Zakarian, Mr Artun    3rd  27.00    male   
1309                         Zakarian, Mr Maprieder    3rd  26.00    male   
1310                               Zenni, Mr Philip    3rd  22.00    male   
1311                               Lievens, Mr Rene    3rd  24.00    male   
1312                                 Zimmerman, Leo    3rd  29.00    male   

      Survived  SexCode  
0            1        1  
1            0        1  
2            0        0  
3            0        1  
4            1        0  
...        ...      ...  
1308         0        0  
1309         0        0  
1310         0        0  
1311         0        0  
1312         0        0  

[1313 rows x 6 columns]

### Grouping Rows by Values

In [52]:
## Group by Passenger Class
dataframe.groupby('PClass').sum()

Age  Survived  SexCode
PClass                            
*          0.00         0        0
1st     8964.92       193      143
2nd     5999.63       119      107
3rd     8016.33       138      212

In [53]:
## Restrict the operation to a specific columns
dataframe[['PClass', 'Age']].groupby('PClass').mean('Age')

Age
PClass           
*             NaN
1st     39.667788
2nd     28.300142
3rd     25.208585

In [54]:
## The Above can also be performed alternatively as below:
dataframe.groupby('PClass')['Age'].mean()

PClass
*            NaN
1st    39.667788
2nd    28.300142
3rd    25.208585
Name: Age, dtype: float64

In [55]:
## Grouping by multiple Columns
dataframe.groupby(['PClass', 'Sex'])[['Age', 'SexCode']].sum()

Age  SexCode
PClass Sex                     
*      male       0.00        0
1st    female  3815.00      143
       male    5149.92        0
2nd    female  2328.00      107
       male    3671.63        0
3rd    female  2323.17      212
       male    5693.16        0

In [56]:
## drop the * row from the data frame
df = dataframe[dataframe['PClass'] != '*']

In [57]:
## Value counts for the df[PClass]
df['PClass'].value_counts()

3rd    711
1st    322
2nd    279
Name: PClass, dtype: int64

### Looping over Columns

In [58]:
## List comprehensions
## return the first five entries after using the list comprehensions
[name.upper() for name in df['Name']][0:4]

['ALLEN, MISS ELISABETH WALTON',
 'ALLISON, MISS HELEN LORAINE',
 'ALLISON, MR HUDSON JOSHUA CREIGHTON',
 'ALLISON, MRS HUDSON JC (BESSIE WALDO DANIELS)']

### Apply function

In [59]:
df['Name'].apply(str.upper)

0                        ALLEN, MISS ELISABETH WALTON
1                         ALLISON, MISS HELEN LORAINE
2                 ALLISON, MR HUDSON JOSHUA CREIGHTON
3       ALLISON, MRS HUDSON JC (BESSIE WALDO DANIELS)
4                       ALLISON, MASTER HUDSON TREVOR
                            ...                      
1308                               ZAKARIAN, MR ARTUN
1309                           ZAKARIAN, MR MAPRIEDER
1310                                 ZENNI, MR PHILIP
1311                                 LIEVENS, MR RENE
1312                                   ZIMMERMAN, LEO
Name: Name, Length: 1312, dtype: object

In [60]:
## Arranging data sorting
print('Option1\n', df.value_counts(['PClass', 'Survived']))

## The above can also be alternatively be performed as follows
print('Option2 for the above\n', df[['PClass', 'Survived']].value_counts())

## You can also implement hierarchical indexing as follows
print('Display results by descending order\n', df[['PClass', 'Survived']].value_counts().sort_index(ascending = False))

## sort the results
print('return the default results\n', df['PClass'].value_counts())

## By default the above sorts the results in descending order
## Removing the default sort
print('Don\'t Sort by default\n', df['PClass'].value_counts(sort= False))

Option1
 PClass  Survived
3rd     0           573
1st     1           193
2nd     0           160
3rd     1           138
1st     0           129
2nd     1           119
dtype: int64
Option2 for the above
 PClass  Survived
3rd     0           573
1st     1           193
2nd     0           160
3rd     1           138
1st     0           129
2nd     1           119
dtype: int64
Display results by descending order
 PClass  Survived
3rd     1           138
        0           573
2nd     1           119
        0           160
1st     1           193
        0           129
dtype: int64
return the default results
 3rd    711
1st    322
2nd    279
Name: PClass, dtype: int64
Don't Sort by default
 1st    322
2nd    279
3rd    711
Name: PClass, dtype: int64


In [61]:
## select the numeric variables
df_cont = df.select_dtypes('float')
df_cont

Age
0     29.00
1      2.00
2     30.00
3     25.00
4      0.92
...     ...
1308  27.00
1309  26.00
1310  22.00
1311  24.00
1312  29.00

[1312 rows x 1 columns]

#### Using *`isin`* to filter data

In [62]:
isin_filter = df['PClass'].isin(['1st', '2nd'])
## Apply the filter
df[isin_filter]

Name PClass    Age     Sex  \
0                     Allen, Miss Elisabeth Walton    1st  29.00  female   
1                      Allison, Miss Helen Loraine    1st   2.00  female   
2              Allison, Mr Hudson Joshua Creighton    1st  30.00    male   
3    Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00  female   
4                    Allison, Master Hudson Trevor    1st   0.92    male   
..                                             ...    ...    ...     ...   
597                          Yrois, Miss Henriette    2nd    NaN  female   
598                  Aldworth, Mr Charles Augustus    2nd  30.00    male   
599                            Brown, Miss Mildred    2nd  24.00  female   
600                                Pernot, Mr Rene    2nd    NaN    male   
601                               Swane, Mr George    2nd  18.00    male   

     Survived  SexCode  
0           1        1  
1           0        1  
2           0        0  
3           0        1  
4           1        0  
..        ...      ...  
597         0        1  
598         0        0  
599         1        1  
600         0        0  
601         0        0  

[601 rows x 6 columns]